### ISSUE TITLE
"AADSTS90023: Invalid STS request" when `acquireToken` is called with `UserNamePasswordParameters` and password contains unicode characters and Seamless Single Sign-On is enabled

### ISSUE THREAD

**Issue by setycz** [AzureAD/microsoft-authentication-library-for-java#580]
Hi guys,

first, this does not seem like a bug in msal4j library, more like a limitation in Azure AD, but I am posting it here in hopes that you can help me validate my observations.

We have a following environment: 
* a user with unicode characters in password ("J123éèàçùµ") in on-prem AD
* users are synced using Azure AD Connect into Azure AD
* Seamless Single Sign-On is enabled (enabled from Azure AD Connect tool)
* our app uses Azure AD for authentication

Reproduction steps:
* call `BaseAzureClient.acquireToken` with `UserNamePasswordParameters` (the passwords is the one user has, with unicode characters: `J123éèàçùµ`)

Result:
* authentication fails
```
com.microsoft.aad.msal4j.MsalServiceException: Server returned error in RSTR - ErrorCode: InvalidRequest. FaultMessage: Invalid Request
    at com.microsoft.aad.msal4j.WSTrustResponse.parse(WSTrustResponse.java:88) ~[msal4j-1.13.3.jar:1.13.3]
    at com.microsoft.aad.msal4j.WSTrustRequest.execute(WSTrustRequest.java:48) ~[msal4j-1.13.3.jar:1.13.3]
    at com.microsoft.aad.msal4j.WSTrustRequest.execute(WSTrustRequest.java:74) ~[msal4j-1.13.3.jar:1.13.3]
    at com.microsoft.aad.msal4j.AcquireTokenByAuthorizationGrantSupplier.processPasswordGrant(AcquireTokenByAuthorizationGrantSupplier.java:98) ~[msal4j-1.13.3.jar:1.13.3]
    at com.microsoft.aad.msal4j.AcquireTokenByAuthorizationGrantSupplier.execute(AcquireTokenByAuthorizationGrantSupplier.java:43) ~[msal4j-1.13.3.jar:1.13.3]
    at com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:69) [msal4j-1.13.3.jar:1.13.3]
    at com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:18) [msal4j-1.13.3.jar:1.13.3]
    at java.util.concurrent.CompletableFuture$AsyncSupply.run$$$capture(CompletableFuture.java:1700) [?:?]
    at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java) [?:?]
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
    at java.lang.Thread.run(Thread.java:834) [?:?]
```

Analysis:
* the call goes into `AcquireTokenByAuthorizationGrantSupplier.processPasswordGrant`
* `userDiscoveryResponse.isAccountFederated()` returns true (because SSO is enabled)
* `WSTrustRequest.execute` contains the right credentials in `body`
* but the Azure responds:
```xml
<?xml version="1.0" encoding="utf-8"?><S:Envelope xmlns:wsa="http://www.w3.org/2005/08/addressing" xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" xmlns:wsp="http://schemas.xmlsoap.org/ws/2004/09/policy" xmlns:wst="http://schemas.xmlsoap.org/ws/2005/02/trust" xmlns:S="http://www.w3.org/2003/05/soap-envelope"><S:Header><psf:pp xmlns:psf="http://schemas.microsoft.com/Passport/SoapServices/SOAPFault"><psf:serverVersion>1</psf:serverVersion><psf:authstate>0x80048800</psf:authstate><psf:reqstatus>0x80045c01</psf:reqstatus><psf:serverInfo ServerTime="2023-01-09T09:09:44.5465943Z">ESTS-PUB-EUS-AZ2-FD038-001.ProdSlices rid:2f945189-49ed-4f6c-942e-5df72da19f02</psf:serverInfo></psf:pp></S:Header><S:Body xmlns:S="http://www.w3.org/2003/05/soap-envelope"><S:Fault><S:Code><S:Value>S:Sender</S:Value><S:Subcode><S:Value>wst:InvalidRequest</S:Value></S:Subcode></S:Code><S:Reason><S:Text xml:lang="en-US">Invalid Request</S:Text></S:Reason><S:Detail><psf:error xmlns:psf="http://schemas.microsoft.com/Passport/SoapServices/SOAPFault"><psf:value>0x80045C01</psf:value><psf:internalerror><psf:code>0x80045C01</psf:code><psf:text>AADSTS90023: Invalid STS request.</psf:text></psf:internalerror></psf:error></S:Detail></S:Fault></S:Body></S:Envelope>
```

We have found out that Azure AD does not allow unicode characters in the password: https://learn.microsoft.com/en-us/azure/active-directory/authentication/concept-sspr-policy#password-policies-that-only-apply-to-cloud-user-accounts , also validation in Azure Portal disallows that. But what is surprising is that when Seamless Single Sign-On is disabled, than authentication works. `userDiscoveryResponse.isAccountFederated()` returns false so there is no call to failing `WSTrustRequest.execute`, could it be that some Azure AD endpoint support passwords with unicode characters and some not?

What blown our mind is that we tried to call `BaseAzureClient.acquireToken` with encoded password using unicode to ASCII converter (https://onlineasciitools.com/convert-unicode-to-ascii), `J123éèàçùµ`  was encoded as `J123Ã©Ã¨Ã Ã§Ã¹Âµ` ,and guess what, the authentication WAS successful. I am not sure what to think about this, but is looks like there is some weird stuff happening in Azure AD when passwords are validated.

Everything seems to indicate, that this is a limitation of Azure AD, would you agree?

**Comment by Avery-Dunn** [comment]
Hello @setycz : Thanks for the thorough investigation, this does seem to either a bug or limitation in Azure AD and probably not something we could fix in our library.

In MSAL Java, we decode some parts of an HTTP response in UTF-8 and encode a specific HTTP request in UTF-8, however the vast majority of the library uses the default encoding for Java strings which is UTF-16.

And regardless that encoding, at no point do we adjust or encode any username/password info, so any characters we get from you should be exactly what we send to Azure AD.

**Comment by siddhijain** [comment]
Closing due to inactivity. Please reopen this or open another one if there are more questions.

**Comment by xRodney** [comment]
Hello,

we have also stumbled upon this limitation and I think I have convincing evidence that this is a bug in this library:

The scenario is exactly as above, except our customer has a password containing a pound sign, let's assume "aaa£".

Digging through the code, and following exactly the same steps as @setycz outlined above, I am able to reconstruct the outgoing HTTP request:

```
POST /$CUSTOMER_DOMAIN/winauth/trust/2005/usernamemixed?client-request-id=f13b7134-fde4-49c4-9e23-fdf9f55b7bca HTTP/1.1
Host: autologon.microsoftazuread-sso.com
return-client-request-id: true
SOAPAction: http://schemas.xmlsoap.org/ws/2005/02/trust/RST/Issue
Content-Type: application/soap+xml; charset=utf-8
Cookie: fpc=Ag537631bkdKqMj96IcqP3xYeBteAQAAAD-etOEOAAAA; stsservicecookie=estsfd; x-ms-gateway-slice=estsfd
Content-Length: 1997

<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope" xmlns:a="http://www.w3.org/2005/08/addressing" xmlns:u="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">
    <s:Header>
        <a:Action s:mustUnderstand="1">http://schemas.xmlsoap.org/ws/2005/02/trust/RST/Issue</a:Action>
        <a:messageID>urn:uuid:4b1d74b2-ea22-4815-bad1-f6a92afe9289</a:messageID>
        <a:ReplyTo>
            <a:Address>http://www.w3.org/2005/08/addressing/anonymous</a:Address>
        </a:ReplyTo>
        <a:To s:mustUnderstand="1">https://autologon.microsoftazuread-sso.com/$CUSTOMER_DOMAIN/winauth/trust/2005/usernamemixed?client-request-id=ef90b0f4-58ef-48a4-8856-d5cba6fd3d83</a:To>
        <o:Security xmlns:o="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" s:mustUnderstand="1">
            <u:Timestamp u:Id="_0">
                <u:Created>2026-06-05T10:38:34Z</u:Created>
                <u:Expires>2026-06-05T10:48:34Z</u:Expires>
            </u:Timestamp>
            <o:UsernameToken u:Id="uuid-256a4f4b-cc14-4576-865e-afdb61bbc238">
                <o:Username>$USERNAME</o:Username>
                <o:Password>aaa£</o:Password>
            </o:UsernameToken>
        </o:Security>
    </s:Header>
    <s:Body>
        <trust:RequestSecurityToken xmlns:trust="http://schemas.xmlsoap.org/ws/2005/02/trust">
            <wsp:AppliesTo xmlns:wsp="http://schemas.xmlsoap.org/ws/2004/09/policy">
                <a:EndpointReference>
                    <a:Address>urn:federation:MicrosoftOnline</a:Address>
                </a:EndpointReference>
            </wsp:AppliesTo>
            <trust:KeyType>http://schemas.xmlsoap.org/ws/2005/05/identity/NoProofKey</trust:KeyType>
            <trust:RequestType>http://schemas.xmlsoap.org/ws/2005/02/trust/Issue</trust:RequestType>
        </trust:RequestSecurityToken>
    </s:Body>
</s:Envelope>
```

($CUSTOMER_DOMAIN and $USERNAME are placeholders to hide the customer data which I am not allowed to publish. I also formatted the XML for better readability. The following steps were performed with the original, uncensored and unformatted data.)

A tell-tale sign is that when I copy-pasted this request to various HTTP clients (Postman, IntelliJ), it went through without issue, indicating a problem with the client library, rather than a server.

Next, I set up my own TCP server to capture the exact bytes going out: `nc -l 9091`.

I then sent the same request from Postman and IntelliJ, and I used a debugger to manipulate the `policy.getUrl()` in WSTrustRequest, so that all requests reached my `http://localhost:9091`.

The result
---------------
In a content header, it is specified that `charset=utf-8`. 

In UTF-8, the [pound sign](https://www.compart.com/en/unicode/U+00A3) is encoded as two bytes `0xC2 0xA3` - and this is exactly what Postman and IntelliJ are sending.

But MSAL sends just one byte `0xA3` - like in ASCII/Latin1.

Why is that?
-----------------
MSAL constructs the whole request by combining Java Strings, converting them to `byte[]` at the very end. Java Strings are stored internally either as UTF16 (exactly 2 bytes per every char) or Latin1 (1 byte per char). My password `aaa£` actually fits into Latin1, so internally is stored on 4 bytes.

MSAL does not do any re-encoding from Java String to UTF-8, instead doing this:

```com.microsoft.aad.msal4j.DefaultHttpClient

private HttpResponse executeHttpPost(HttpRequest httpRequest) throws Exception {

        final HttpURLConnection conn = openConnection(httpRequest.url());
        configureAdditionalHeaders(conn, httpRequest);
        conn.setRequestMethod("POST");
        conn.setDoOutput(true);

        DataOutputStream wr = null;
        try {
            wr = new DataOutputStream(conn.getOutputStream());
            wr.writeBytes(httpRequest.body());           //   body() returns a String
            wr.flush();

            return readResponseFromConnection(conn);
        } finally {
            if (wr != null) {
                wr.close();
            }
        }
    }
```

delegating to Java's DataOutputStream: 

```java.io.DataOutputStream
    public final void writeBytes(String s) throws IOException {
        int len = s.length();
        for (int i = 0 ; i < len ; i++) {
            out.write((byte)s.charAt(i));
        }
        incCount(len);
    }

```

... which simply skips the potential higher byte, clamping each character to just one byte.

Note that a pound sign in particular fits into a single byte, but UTF-8 still requires it to be prefixed with 0xC2 to denote a two byte sequence.


TLDR
-------

MSAL, which is supposed to use UTF-8, sends requests in Latin1, causing them to be rightfully rejected as invalid by the server.

Please, consider reopening this issue and fixing it. Thanks.
